src/structured_data.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright 2015 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under (1) the MaidSafe.net Commercial License,
+// version 1.0 or later, or (2) The General Public License (GPL), version 3, depending on which
+// licence you accepted on initial access to the Software (the "Licences").
+//
+// By contributing code to the SAFE Network Software, or to this project generally, you agree to be
+// bound by the terms of the MaidSafe Contributor Agreement, version 1.0.  This, along with the
+// Licenses can be found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+use rustc_serialize::{Encodable, Encoder, Decoder};
+use cbor;
+use error::RoutingError;
+use NameType;
+use sodiumoxide::crypto;
+use std::str;
+
+/// StructuredData
+/// These types may be stored unsigned with previous and current owner keys
+/// set to the same keys. Updates though require a signature to validate
+#[derive(Clone, RustcDecodable, RustcEncodable)]
+pub struct StructuredData {
```

Comment:
Cheers @ustulation 

Thoughts 

1. Agreed 

2. Agreed (although on the fence as to whether NameType should exist at all and instead not be actually a hash, which I think it always is) Should be note sodium seems to not like copying hashes (timing attacks). So this needs security review 

3. Agreed 

4. Not sure, @inetic made the point a vector by default provides the ability to be zero length, which does make sense as it's implicit in the type. (it used to be an Option)

5. Again not sure, if your prev owners field is empty the signature check would fail would it not ? So an invariant is prev owners must have signed data over to new owner. 

And lastly PR will be welcomed by @inetic I am sure :-)